src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,186 @@
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        final Node<T> l = last;
+        Node<T> newNode = new Node<T>(l, value, null);
+        last = newNode;
+        l.next = newNode;
+        size++;
+    }
+
+    private void addInMiddle(T value, Node<T> successor) {
+        Node<T> predecessor = successor.prev;
+        Node<T> newNode = new Node<>(predecessor, value, successor);
+        successor.prev = newNode;
+        if (predecessor == null) {
+            first = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexNotEqualsSize(index);
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexNotEqualsSize(index);
+        T oldValue = findNode(index).item;
+        findNode(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexNotEqualsSize(index);
+        final T oldValue = findNode(index).item;
+        Node<T> nextFromRemoved = findNode(index).next;
+        Node<T> prevFromRemoved = findNode(index).prev;
+        if (last == findNode(index) && size != 1) {
+            prevFromRemoved.next = null;
+            last = prevFromRemoved;
+        } else if (first == findNode(index) && size != 1) {
+            nextFromRemoved.prev = null;
+            first = nextFromRemoved;
+        } else if (size != 1) {
+            nextFromRemoved.prev = prevFromRemoved;
+            prevFromRemoved.next = nextFromRemoved;
+        }
+        size--;
+        sizeChecker();
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i <= size; i++) {
+            if (findNode(i).item == object || findNode(i).item != null
+                    && findNode(i).item.equals(object)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
+    private void sizeChecker() {
+        if (size == 0) {
+            first = null;
+            last = null;
+        }
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isFirst() {
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index + ", with size: " + size);
+        }
+    }
+
+    private void checkIndexNotEqualsSize(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index + ", with size: " + size);
+        }
+    }
+
+    private Node<T> findNode(int index) {
```

Comment:
Also bad naming inside this method. Fix it
```suggestion
    private Node<T> getNodeByIndex(int index) {
```